File: apps/mobile/lib/core/scroll_engine.dart
import 'package:flutter/scheduler.dart';

/// Time-elapsed based auto-scroll engine — mirrors the web app's
/// [ScrollEngine] (apps/teleprompt/src/lib/scroll-engine.ts). Position is
/// derived from wall-clock delta on every tick rather than a fixed per-frame
/// increment, so playback stays smooth regardless of device frame rate.
class ScrollEngine {
  ScrollEngine({
    required TickerProvider vsync,
    required this.onUpdate,
    this.onEnd,
    double baseRate = 40,
  }) : _baseRate = baseRate {
    _ticker = vsync.createTicker(_onTick);
  }

  final void Function(double position) onUpdate;
  final void Function()? onEnd;

  late final Ticker _ticker;
  double _position = 0;
  double _speed = 1;
  double _baseRate;
  double _maxPosition = double.infinity;
  Duration _lastElapsed = Duration.zero;

  double get position => _position;
  bool get isPlaying => _ticker.isTicking;
  double get baseRate => _baseRate;
  double get speed => _speed;
  double get maxPosition => _maxPosition;

  void setBaseRate(double rate) => _baseRate = rate;
  void setSpeed(double speed) => _speed = speed;
  void setMaxPosition(double max) => _maxPosition = max;

  void seek(double position) {
    _position = position.clamp(0, _maxPosition == double.infinity ? position : _maxPosition);
    onUpdate(_position);
  }

  void nudge(double deltaPx) => seek(_position + deltaPx);

  void play() {
    if (_ticker.isTicking) return;
    _lastElapsed = Duration.zero;
    _ticker.start();
  }

  void pause() {
    if (_ticker.isTicking) _ticker.stop();
  }

  void toggle() {
    if (_ticker.isTicking) {
      pause();
    } else {
      play();
    }
  }

  void dispose() {
    _ticker.dispose();
  }

  void _onTick(Duration elapsed) {
    final dtSeconds = _lastElapsed == Duration.zero
        ? 0.0
        : (elapsed - _lastElapsed).inMicroseconds / 1e6;
    _lastElapsed = elapsed;

    final next = _position + _baseRate * _speed * dtSeconds;
    if (next >= _maxPosition) {
      _position = _maxPosition;
      onUpdate(_position);
      pause();
      onEnd?.call();
      return;
    }
    _position = next;
    onUpdate(_position);
  }
}

File: apps/teleprompt/src/lib/scroll-engine.ts
/**
 * Time-elapsed based auto-scroll engine (cahier des charges §12).
 * Position is derived from wall-clock delta each animation frame, not from a
 * fixed per-tick increment, so playback stays smooth regardless of the
 * device's actual frame rate and never accumulates drift while paused.
 */
export type ScrollEngineListener = (position: number) => void;

export interface ScrollEngineOptions {
  /** px/second at speed = 1x */
  baseRate: number;
  onUpdate: ScrollEngineListener;
  /** Called when position reaches maxPosition during playback */
  onEnd?: () => void;
}

export class ScrollEngine {
  private position = 0;
  private speed = 1;
  private baseRate: number;
  private playing = false;
  private rafId: number | null = null;
  private lastTimestamp = 0;
  private maxPosition = Infinity;
  private onUpdate: ScrollEngineListener;
  private onEnd?: () => void;

  constructor(options: ScrollEngineOptions) {
    this.baseRate = options.baseRate;
    this.onUpdate = options.onUpdate;
    this.onEnd = options.onEnd;
  }

  setBaseRate(rate: number) {
    this.baseRate = rate;
  }

  getBaseRate() {
    return this.baseRate;
  }

  setSpeed(speed: number) {
    this.speed = speed;
  }

  getSpeed() {
    return this.speed;
  }

  setMaxPosition(max: number) {
    this.maxPosition = max;
  }

  getMaxPosition() {
    return this.maxPosition;
  }

  getPosition() {
    return this.position;
  }

  isPlaying() {
    return this.playing;
  }

  seek(position: number) {
    this.position = Math.max(0, Math.min(position, this.maxPosition));
    this.onUpdate(this.position);
  }

  nudge(deltaPx: number) {
    this.seek(this.position + deltaPx);
  }

  play() {
    if (this.playing) return;
    this.playing = true;
    this.lastTimestamp = performance.now();
    this.tick();
  }

  pause() {
    this.playing = false;
    if (this.rafId !== null) {
      cancelAnimationFrame(this.rafId);
      this.rafId = null;
    }
  }

  toggle() {
    if (this.playing) this.pause();
    else this.play();
  }

  dispose() {
    this.pause();
  }

  private tick = () => {
    if (!this.playing) return;
    const now = performance.now();
    const dtSeconds = (now - this.lastTimestamp) / 1000;
    this.lastTimestamp = now;

    const next = this.position + this.baseRate * this.speed * dtSeconds;
    if (next >= this.maxPosition) {
      this.position = this.maxPosition;
      this.onUpdate(this.position);
      this.pause();
      this.onEnd?.();
      return;
    }

    this.position = next;
    this.onUpdate(this.position);
    this.rafId = requestAnimationFrame(this.tick);
  };
}
